KAFKA-2476: Add Decimal, Date, and Timestamp logical types.

ewencp opened the pull request:
To support Decimal, this also adds support for schema parameters, which is an
extra set of String key value pairs which provide extra information about the
schema. For Decimal, this is used to encode the scale parameter, which is part
of the schema instead of being passed with every value.

ewencp committed:
KAFKA-2476: Add Decimal, Date, and Timestamp logical types.

To support Decimal, this also adds support for schema parameters, which is an
extra set of String key value pairs which provide extra information about the
schema. For Decimal, this is used to encode the scale parameter, which is part
of the schema instead of being passed with every value.

guozhangwang commented on copycat/api/src/test/java/org/apache/kafka/copycat/data/CopycatSchemaTest.java:
```diff
@@ -208,14 +208,15 @@ public void testValidateValueMismatchStructWrongNestedSchema() {
-        CopycatSchema differentName = new CopycatSchema(Schema.Type.INT8, false, null, "otherName", 2, "doc", null, null, null);
-        CopycatSchema differentVersion = new CopycatSchema(Schema.Type.INT8, false, null, "name", 4, "doc", null, null, null);
-        CopycatSchema differentDoc = new CopycatSchema(Schema.Type.INT8, false, null, "name", 2, "other doc", null, null, null);
+        CopycatSchema s1 = new CopycatSchema(Schema.Type.INT8, false, null, "name", 2, "doc", null, null, null, null);
```
Could we have additional CopycatSchema constructors that put default field values as null?

ewencp replied:
We definitely could. I expected that people would almost always use `SchemaBuilder` to construct `Schema` objects so I wasn't that worried about making a variety of constructors available. Aside from tests, we only directly construct these objects in a one or two locations.

guozhangwang commented on copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaBuilderTest.java:
```diff
@@ -188,6 +190,21 @@ public void testBytesBuilderInvalidDefault() {
+        assertTypeAndDefault(schema, Schema.Type.STRING, false, null);
+        assertMetadata(schema, null, null, null, expectedParameters);
+    }
+
```
Add another test case for invalidating parameter values as non-strings?

ewencp replied:
Not sure I understand. The only two ways to pass in parameters are via `SchemaBuilder parameter(String propertyName, String propertyValue)` and `SchemaBuilder parameters(Map<String, String> props)`. Doesn't that already ensure non-string values won't be passed in?

guozhangwang replied:
You are right, I was over-concerned.

ewencp replied:
Appropriately concerned. I started writing a test before I realized I couldn't get a non-String value in there :)

guozhangwang commented on copycat/json/src/main/java/org/apache/kafka/copycat/json/JsonConverter.java:
```diff
@@ -443,6 +527,12 @@ private static JsonNode convertToJson(Schema schema, Object value) {
+        if (schema != null && schema.name() != null) {
+            LogicalTypeConverter logicalConverter = TO_JSON_LOGICAL_CONVERTERS.get(schema.name());
+            if (logicalConverter != null)
+                value = logicalConverter.convert(schema, value);
```
Do we want to overwrite the passed-in value reference here?

guozhangwang commented:
LGTM except minor comments. Could you rebase?

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into kafka-2476-copycat-logical-types

ewencp committed:
Address review comments. Add shorthand CopycatSchema constructors and don't overwrite a method parameter's value in JsonConverter.

ewencp commented:
@guozhangwang merged and addressed most of your comments. The one I wasn't sure about was the one about invalidating parameter values as non-strings. Can you clarify what you were looking for there?
